lib/portlayer/exec/commit.go

```diff
@@ -0,0 +1,210 @@
+			})
+		}
+
+		if err != nil {
+			log.Errorf("Something failed. Spec was %+v", *h.Spec.Spec())
+			return err
+		}
+
+		h.vm = vm.NewVirtualMachine(ctx, sess, res.Result.(types.ManagedObjectReference))
+		c = newContainer(&h.containerBase)
+		Containers.Put(c)
+		// inform of creation irrespective of remaining operations
+		publishContainerEvent(c.ExecConfig.ID, time.Now().UTC(), events.ContainerCreated)
+
+		// clear the spec as we've acted on it - this prevents a reconfigure from occurring in follow-on
+		// processing
+		h.Spec = nil
+	}
+
+	// if we're stopping the VM, do so before the reconfigure to preserve the extraconfig
+	refresh := true
+	if h.TargetState() == StateStopped {
+		if h.Runtime == nil {
+			log.Warnf("Commit called with incomplete runtime state for %s", h.ExecConfig.ID)
+		}
+
+		if h.Runtime != nil && h.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOff {
+			log.Infof("Dropping duplicate power off operation for %s", h.ExecConfig.ID)
+		} else {
+			// stop the container
+			if err := c.stop(ctx, waitTime); err != nil {
+				return err
+			}
+
+			// inform of creation irrespective of remaining operations
+			publishContainerEvent(h.ExecConfig.ID, time.Now().UTC(), events.ContainerStopped)
+
+			// we must refresh now to get the new ChangeVersion - this is used to gate on powerstate in the reconfigure
+			// because we cannot set the ExtraConfig if the VM is powered on. There is still a race here unfortunately because
+			// tasks don't appear to contain the new ChangeVersion
+			// we don't use refresh because we want to keep the extraconfig state
+			base, err := h.updates(ctx)
+			if err != nil {
+				// TODO: can we recover here, or at least set useful state for inspection?
+				return err
+			}
+			h.Runtime = base.Runtime
+			h.Config = base.Config
+
+			refresh = false
+		}
+	}
+
+	// reconfigure operation
+	if h.Spec != nil {
+		if h.Runtime == nil {
+			log.Errorf("Refusing to perform reconfigure operation with incomplete runtime state for %s", h.ExecConfig.ID)
+		} else {
+			// ensure that our logic based on Runtime state remains valid
+
+			// NOTE: this inline refresh can be removed when switching away from guestinfo where we have non-persistence issues
+			// when updating ExtraConfig via the API with a powered on VM - we therefore have to be absolutely certain about the
+			// power state to decide if we can continue without nilifying extraconfig
+
+			for s := h.Spec.Spec(); ; refresh, s = true, h.Spec.Spec() {
+				// FIXME!!! this is a temporary hack until the concurrent modification retry logic is in place
+				if refresh {
+					base, err := h.updates(ctx)
+					if err == nil {
+						h.Runtime = base.Runtime
+						h.Config = base.Config
+					}
+				}
+
+				s.ChangeVersion = h.Config.ChangeVersion
+
+				// nilify ExtraConfig if vm is running
+				if h.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
```

Comment:
hm... ok didn't see that `s` is "refreshed" for every iteration of the loop. nvm.